File: backend/app/services/humanize_answers.py
from __future__ import annotations

from typing import Any, TypedDict


class HumanField(TypedDict, total=False):
    key: str
    label: str
    kind: str
    value_text: str
    raw_value: Any


def _human_fallback_label(field: str, ftype: str | None = None) -> str:
    key = (field or "").lower()
    t = (ftype or "").lower()
    if t == "email" or "email" in key:
        return "Email"
    if t == "phone" or "phone" in key or "tel" in key:
        return "Телефон"
    if key == "name" or "first_name" in key:
        return "Имя"
    if "comment" in key:
        return "Комментарий"
    if "rating" in key or "nps" in key:
        return "Оценка"
    return field


def _safe_str(v: Any) -> str:
    if v is None:
        return ""
    if isinstance(v, (str, int, float, bool)):
        return str(v)
    try:
        import json

        return json.dumps(v, ensure_ascii=False)
    except Exception:
        return str(v)


def humanize_extra_answers(
    *,
    schema: dict[str, Any] | None,
    answers: dict[str, Any] | None,
    dev: bool = False,
    hide_unknown: bool = True,
    skip_keys: set[str] | None = None,
) -> list[HumanField]:
    """Convert raw answers into UI-friendly rows.

    Mirrors FE mapping (rating/text/choice/contact + __other_text).
    Used to avoid returning full raw JSON answers/meta in non-dev mode.
    """
    schema = schema or {}
    answers = answers or {}
    skip = skip_keys or set()

    defs: dict[str, dict[str, Any]] = {}
    slides = schema.get("slides")
    if not isinstance(slides, list):
        slides = []

    for slide in slides:
        if not isinstance(slide, dict):
            continue
        stype = str(slide.get("type") or "")
        title = str(slide.get("title") or "").strip()

        if stype in ("rating", "nps"):
            field = str(slide.get("field") or "rating_overall")
            defs[field] = {
                "label": title or _human_fallback_label(field, stype),
                "kind": "rating",
            }
            continue

        if stype == "text":
            field = str(slide.get("field") or "").strip()
            if not field:
                continue
            defs[field] = {
                "label": title or _human_fallback_label(field, stype),
                "kind": "text",
            }
            continue

        if stype == "choice":
            field = str(slide.get("field") or "").strip()
            if not field:
                continue
            mode = str(slide.get("mode") or "single")
            raw_options = slide.get("options")
            options: list[dict[str, str]] = []
            if isinstance(raw_options, list):
                for o in raw_options:
                    if not isinstance(o, dict):
                        continue
                    val = str(o.get("value") or "").strip()
                    lab = str(o.get("label") or o.get("value") or "").strip()
                    if val and lab:
                        options.append({"value": val, "label": lab})

            defs[field] = {
                "label": title or _human_fallback_label(field, stype),
                "kind": "choice_multi" if mode == "multi" else "choice_single",
                "options": options,
            }
            continue

        if stype == "contact":
            fields = slide.get("fields")
            if not isinstance(fields, list):
                fields = []
            for f in fields:
                if not isinstance(f, dict):
                    continue
                field = str(f.get("field") or "").strip()
                if not field:
                    continue
                ftype = str(f.get("type") or "").strip()
                flabel = str(f.get("label") or "").strip()
                defs[field] = {
                    "label": flabel or _human_fallback_label(field, ftype),
                    "kind": "contact",
                }

    other_text: dict[str, str] = {}
    for k, v in list(answers.items()):
        if isinstance(k, str) and k.endswith("__other_text"):
            base = k[: -len("__other_text")]
            txt = str(v or "").strip()
            if txt:
                other_text[base] = txt

    rows: list[HumanField] = []

    for k, raw in list(answers.items()):
        if raw is None:
            continue
        if not isinstance(k, str):
            continue
        if k.endswith("__other_text"):
            continue
        if k in skip:
            continue

        d = defs.get(k)
        if d is None:
            if hide_unknown:
                continue
            row: HumanField = {
                "key": k,
                "label": f"Неизвестное поле ({k})",
                "kind": "unknown",
                "value_text": _safe_str(raw),
            }
            if dev:
                row["raw_value"] = raw
            rows.append(row)
            continue

        kind = str(d.get("kind") or "unknown")
        label = str(d.get("label") or k)
        value_text = ""

        if kind in ("rating", "text", "contact"):
            value_text = _safe_str(raw)
        elif kind == "choice_single":
            rv = _safe_str(raw)
            if rv == "other":
                value_text = f"Другое: {other_text.get(k)}" if other_text.get(k) else "Другое"
            else:
                opt = next((o for o in d.get("options") or [] if o.get("value") == rv), None)
                value_text = (opt.get("label") if opt else None) or rv
        elif kind == "choice_multi":
            if isinstance(raw, list):
                arr = [_safe_str(x).strip() for x in raw]
                arr = [x for x in arr if x]
            else:
                rv = _safe_str(raw).strip()
                arr = [rv] if rv else []

            parts: list[str] = []
            for rv in arr:
                if rv == "other":
                    parts.append(f"Другое: {other_text.get(k)}" if other_text.get(k) else "Другое")
                    continue
                opt = next((o for o in d.get("options") or [] if o.get("value") == rv), None)
                parts.append((opt.get("label") if opt else None) or rv)
            value_text = ", ".join([p for p in parts if p.strip()])

        if (not dev) and (not str(value_text).strip()):
            continue

        row2: HumanField = {
            "key": k,
            "label": label,
            "kind": kind,
            "value_text": value_text,
        }
        if dev:
            row2["raw_value"] = raw
        rows.append(row2)

    return rows

File: frontend/src/shared/humanizeAnswers.ts
type ChoiceOption = { value: string; label: string };

type FieldDef = {
  key: string;
  label: string;
  kind: "rating" | "text" | "choice_single" | "choice_multi" | "contact";
  options?: ChoiceOption[];
};

export type HumanFieldRow = {
  key: string;
  label: string;
  kind: FieldDef["kind"] | "unknown";
  valueText: string;
  rawValue: unknown;
};

function humanFallbackLabel(field: string, ftype?: string) {
  const key = (field || "").toLowerCase();
  const t = (ftype || "").toLowerCase();
  if (t === "email" || key.includes("email")) return "Email";
  if (t === "phone" || key.includes("phone") || key.includes("tel")) return "Телефон";
  if (key === "name" || key.includes("first_name")) return "Имя";
  if (key.includes("comment")) return "Комментарий";
  if (key.includes("rating") || key.includes("nps")) return "Оценка";
  return field;
}

function safeStr(v: unknown) {
  if (v == null) return "";
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "boolean") return String(v);
  try {
    return JSON.stringify(v);
  } catch {
    return String(v);
  }
}

function buildDefsFromSchema(schema: any): Record<string, FieldDef> {
  const defs: Record<string, FieldDef> = {};
  const slides = Array.isArray(schema?.slides) ? schema.slides : [];

  for (const slide of slides) {
    if (!slide || typeof slide !== "object") continue;
    const stype = String((slide as any).type ?? "");
    const title = String((slide as any).title ?? "").trim();

    if (stype === "rating" || stype === "nps") {
      const field = String((slide as any).field ?? "rating_overall");
      defs[field] = {
        key: field,
        label: title || humanFallbackLabel(field, stype),
        kind: "rating",
      };
      continue;
    }

    if (stype === "text") {
      const field = String((slide as any).field ?? "");
      if (!field) continue;
      defs[field] = {
        key: field,
        label: title || humanFallbackLabel(field, stype),
        kind: "text",
      };
      continue;
    }

    if (stype === "choice") {
      const field = String((slide as any).field ?? "");
      if (!field) continue;
      const mode = String((slide as any).mode ?? "single");
      const rawOptions = (slide as any).options;
      const options: ChoiceOption[] = Array.isArray(rawOptions)
        ? rawOptions
            .filter((o: any) => o && typeof o === "object")
            .map((o: any) => ({
              value: String(o.value ?? ""),
              label: String(o.label ?? o.value ?? ""),
            }))
            .filter((o: any) => o.value && o.label)
        : [];
      defs[field] = {
        key: field,
        label: title || humanFallbackLabel(field, stype),
        kind: mode === "multi" ? "choice_multi" : "choice_single",
        options,
      };
      continue;
    }

    if (stype === "contact") {
      const fields = Array.isArray((slide as any).fields) ? (slide as any).fields : [];
      for (const f of fields) {
        if (!f || typeof f !== "object") continue;
        const field = String((f as any).field ?? "");
        if (!field) continue;
        const ftype = String((f as any).type ?? "");
        const flabel = String((f as any).label ?? "").trim();
        defs[field] = {
          key: field,
          label: flabel || humanFallbackLabel(field, ftype),
          kind: "contact",
        };
      }
    }
  }

  return defs;
}

function kindLabel(kind: HumanFieldRow["kind"]) {
  switch (kind) {
    case "rating":
      return "Оценка";
    case "text":
      return "Текст";
    case "choice_single":
      return "Выбор (один)";
    case "choice_multi":
      return "Выбор (несколько)";
    case "contact":
      return "Контакт";
    default:
      return "Поле";
  }
}

export function humanizeAnswers(
  schema: any,
  answers: Record<string, unknown>,
  opts?: { dev?: boolean; hideUnknown?: boolean }
): HumanFieldRow[] {
  const dev = Boolean(opts?.dev);
  const hideUnknown = opts?.hideUnknown ?? !dev;

  const defs = buildDefsFromSchema(schema);
  const otherText: Record<string, string> = {};

  // Собираем __other_text заранее и не отображаем его отдельной строкой.
  for (const [k, v] of Object.entries(answers || {})) {
    if (k.endsWith("__other_text")) {
      const base = k.slice(0, -"__other_text".length);
      const txt = String(v ?? "").trim();
      if (txt) otherText[base] = txt;
    }
  }

  const rows: HumanFieldRow[] = [];

  for (const [k, raw] of Object.entries(answers || {})) {
    if (raw == null) continue;
    if (k.endsWith("__other_text")) continue;

    const def = defs[k];
    if (!def) {
      if (hideUnknown) continue;
      rows.push({
        key: k,
        label: `Неизвестное поле (${k})`,
        kind: "unknown",
        valueText: safeStr(raw),
        rawValue: raw,
      });
      continue;
    }

    let valueText = "";

    if (def.kind === "rating" || def.kind === "text" || def.kind === "contact") {
      valueText = safeStr(raw);
    }

    if (def.kind === "choice_single") {
      const rv = safeStr(raw);
      if (rv === "other") {
        valueText = otherText[k] ? `Другое: ${otherText[k]}` : "Другое";
      } else {
        const opt = (def.options || []).find((o) => o.value === rv);
        valueText = opt?.label ?? rv;
      }
    }

    if (def.kind === "choice_multi") {
      const arr = Array.isArray(raw) ? raw.map((x) => safeStr(x)).filter(Boolean) : [safeStr(raw)].filter(Boolean);
      const parts: string[] = [];
      for (const rv of arr) {
        if (rv === "other") {
          parts.push(otherText[k] ? `Другое: ${otherText[k]}` : "Другое");
          continue;
        }
        const opt = (def.options || []).find((o) => o.value === rv);
        parts.push(opt?.label ?? rv);
      }
      valueText = parts.join(", ");
    }

    if (!dev && String(valueText).trim() === "") continue;

    rows.push({
      key: k,
      label: def.label,
      kind: def.kind,
      valueText,
      rawValue: raw,
    });
  }

  return rows;
}

export function humanKindLabel(kind: HumanFieldRow["kind"]) {
  return kindLabel(kind);
}
